File: prototype/solvers/crank_nicolson.py
import numpy as np
import models.option

def tridiagonal_thomas(lower, main, upper, rhs):
    """
    Perform a Thomas algorithm that maintains the original coefficients
    """
    size = len(rhs)
    x = np.zeros(size,dtype='float64')
    y = np.zeros(size,dtype='float64')
    ans = np.zeros(size,dtype='float64')

    x[0] = upper[0] / main[0]
    y[0] = rhs[0] / main[0]

    for i in range(1, size):
        denom = main[i] - lower[i] * x[i - 1]
        x[i] = upper[i] / denom
        y[i] = (rhs[i] - lower[i] * y[i - 1]) / denom

    ans[size - 1] = y[size - 1]

    for i in reversed(range(size - 1)):
        ans[i] = y[i] - x[i] * ans[i + 1]
    
    return ans

def solve_crank_nicolson(option, S_max, T, N, J, V, S, t):
    """
    We take finite differences of our PDE with f approximating the spatial part of the equation

    (V_new[i] - V_old[i]) / dt = 0.5(f(V_old[i]) + f(V_new[i]))

    We approximate f with a tridiagonal matrix A such that f(v) = -Av

    (I - 0.5dtA)V_new = (I + 0.5dtA)V_old

    We call these new matrices ML and MR
    """
    
    sigma = option.sigma
    r = option.r
    dS = S_max / J
    dt = T / N

    # S values for interior points
    S_interior = S[1:J]

    # lower, main, upper diagonal of A multiplied by dt
    a = 0.5 * sigma**2 * S_interior**2 * (dt / dS**2) - 0.5 * r * S_interior * (dt / dS)
    b = -sigma**2 * S_interior**2 * (dt / dS**2) - r * dt
    c = 0.5 * sigma**2 * S_interior**2 * (dt / dS**2) + 0.5 * r * S_interior * (dt / dS)

    # lower, main, upper diagonal of ML and MR
    ML_lower = -0.5 * a
    ML_main = 1 - 0.5 * b
    ML_upper = -0.5 * c

    MR_lower = 0.5 * a
    MR_main = 1 + 0.5 * b
    MR_upper = 0.5 * c

    for n in reversed(range(N)):
        # Set boundary conditions for the current time step
        option.option_price_boundary(V[n, :], S, t[n])

        rhs = np.zeros(J - 1)
        # handle pure interior calculations
        for idx, j in enumerate(range(1, J)):
            rhs[idx] = (
                MR_lower[idx] * V[n + 1, j - 1] +
                MR_main[idx] * V[n + 1, j] +
                MR_upper[idx] * V[n + 1, j + 1]
            )

        # Adjust rhs for boundary conditions
        rhs[0] -= ML_lower[0] * V[n, 0]
        rhs[-1] -= ML_upper[-1] * V[n, -1]

        V[n, 1:J] = tridiagonal_thomas(ML_lower, ML_main, ML_upper, rhs)
        
        # Apply early exercise condition for American options
        option.early_exercise_condition(V[n, :], S, t[n])

    # set boundaries for zero
    option.option_price_boundary(V[0, :], S, t[0])
    
    # Apply final early exercise condition
    option.early_exercise_condition(V[0, :], S, t[0])

    return V

File: cpp/solvers/crank_nicolson.cpp
#include <vector>
#include "../models/option.h"

std::vector<double> tridiagonal_thomas(
    const std::vector<double>& lower,
    const std::vector<double>& main,
    const std::vector<double>& upper,
    const std::vector<double>& rhs
) {
    size_t size = main.size();
    std::vector<double> x(size, 0.0);
    std::vector<double> y(size, 0.0);
    std::vector<double> ans(size, 0.0);

    x[0] = upper[0] / main[0];
    y[0] = rhs[0] / main[0];

    for (size_t i = 1; i < size; ++i) {
        double denom = main[i] - lower[i] * x[i - 1];
        x[i] = upper[i] / denom;
        y[i] = (rhs[i] - lower[i] * y[i - 1]) / denom;
    }

    ans[size - 1] = y[size - 1];

    for (int i = size - 2; i >= 0; --i) {
        ans[i] = y[i] - x[i] * ans[i + 1];
    }

    return ans;
}

double* solve_crank_nicolson(
    const Option& option,
    const double S_max,
    const double T,
    const int N,
    const int J,
    double* V,
    const double* S,
    const double* t
) {
    const double sigma = option.getSigma();
    const double r = option.getR();
    const double dS = S_max / J;
    const double dt = T / N;

    std::vector<double> a(J - 1);
    std::vector<double> b(J - 1);
    std::vector<double> c(J - 1);
    std::vector<double> ML_lower(J - 1);
    std::vector<double> ML_main(J - 1);
    std::vector<double> ML_upper(J - 1);
    std::vector<double> MR_lower(J - 1);
    std::vector<double> MR_main(J - 1);
    std::vector<double> MR_upper(J - 1);
    std::vector<double> rhs(J - 1);
    std::vector<double> thomas_vector(J - 1);

    double sq_sigma = sigma * sigma;
    double sq_S = 0.0;
    double dt_over_sq_dS = (dt / (dS * dS));
    for (int j = 1; j < J; ++j) {
        sq_S = (j * dS) * (j * dS);
        a[j - 1] = 0.5 * sq_sigma * sq_S * dt_over_sq_dS - 0.5 * r * j * dt;
        b[j - 1] = -sq_sigma * sq_S * dt_over_sq_dS - r * dt;
        c[j - 1] = 0.5 * sq_sigma * sq_S * dt_over_sq_dS + 0.5 * r * j * dt;

        ML_lower[j - 1] = -0.5 * a[j - 1];
        ML_main[j - 1] = 1 - 0.5 * b[j - 1];
        ML_upper[j - 1] = -0.5 * c[j - 1];

        MR_lower[j - 1] = 0.5 * a[j - 1];
        MR_main[j - 1] = 1 + 0.5 * b[j - 1];
        MR_upper[j - 1] = 0.5 * c[j - 1];

        rhs[j - 1] = 0.0;
    }

    for (int n = N - 1; n > -1; n--) {
        option.option_price_boundary(V + n * (J + 1), S, t[n], J + 1);
        
        for (int j = 0; j < J - 1; j++) {
            rhs[j] =
                MR_lower[j] * *(V + (n + 1) * (J + 1) + j) +
                MR_main[j] * *(V + (n + 1) * (J + 1) + j + 1) +
                MR_upper[j] * *(V + (n + 1) * (J + 1) + j + 2);
        }

        rhs[0] -= ML_lower[0] * *(V + n * (J + 1));
        rhs[J - 2] -= ML_upper[J - 2] * *(V + n * (J + 1) + J);

        thomas_vector = tridiagonal_thomas(ML_lower, ML_main, ML_upper, rhs);
        for (int j = 1; j < J; j++) {
            *(V + n * (J + 1) + j) = thomas_vector[j - 1];
        }

        // Apply early exercise condition for American options after solving for the time step
        option.early_exercise_condition(V + n * (J + 1), S, t[n], J + 1);
    }

    option.option_price_boundary(V, S, t[0], J + 1);

    return V;
}